MetaMask/metamask-extension
# Add support for WalletConnect in hardware wallets

## Explanation

we have recently developped [eth-walletconnect-keyring](https://github.com/Toporin/eth-walletconnect-keyring), an implementation of MetaMask's Keyring interface that uses the WalletConnect protocol to connect a (hardware) wallet to Metamask. A beta release of Metamask with this solution is available here for testing:
https://github.com/Toporin/metamask-extension/releases/tag/v10.13.0-WC

This is an early proof-of-concept, but it would allow any wallet (hardware/software) that supports the WalletConnect protocol to connect to Metamask. It is a standard and neutral solution, quite simple and without USB/hardware requirements, so we hope that the Metamask team would be more willing to integrate it (like the QR-code approach supported by multiple vendors).

Our first motivation is to allow users of the [Satochip hardware wallet](https://satochip.io) to use Metamask natively, but this generic approach could be used by any vendor such as BitBox02, CoolWalletS, D'Cent or mobile phone wallets by simply implementing WalletConnect.

Here is a [short video demo](https://www.youtube.com/watch?v=IbAvv5XyUVU) showing how to connect a wallet to Metamask through the eth-walletconnect-keyring plugin. When using the WalletConnect option, a new tab opens with a react app that will handle the actual WalletConect connection. In this demo, we used a local server (for development only), in production, the react app is currently hosted on [github-page](https://github.com/Toporin/WalletConnect-Bridge).

Everything is open source, questions, comments and suggestions are welcome!

## Screenshots/Screencaps

### Before



### After

![Screenshot from 2022-06-08 09-53-29](https://user-images.githubusercontent.com/10826856/172581028-5847da45-1428-4df7-b8b9-0a3f49478788.png)
![Screenshot from 2022-06-08 09-58-38](https://user-images.githubusercontent.com/10826856/172581032-a42a5032-789c-49b2-a38a-08f39c86516e.png)
![Screenshot from 2022-06-08 09-58-53](https://user-images.githubusercontent.com/10826856/172581036-b867e251-82ff-4887-a16e-97efcd571cba.png)
![Screenshot from 2022-06-08 09-59-18](https://user-images.githubusercontent.com/10826856/172581040-653af6eb-5ebe-4c90-9d8d-f2f2c164d62a.png)
![Screenshot from 2022-06-08 09-59-49](https://user-images.githubusercontent.com/10826856/172581042-4bbd98a5-cb3b-409d-8a02-379160a10141.png)
![Screenshot from 2022-06-08 10-00-06](https://user-images.githubusercontent.com/10826856/172581048-508a41ba-68ea-4ce6-b0c5-ace077df1327.png)

## Manual Testing Steps

- In the menu, click on "Connect to hardware wallet"
- In the options, click on the "WalletConnect" button
- A webpage to "https://toporin.github.io/WalletConnect-Bridge/" will open in a new tab: click on the "Connect to WalletConnect" button
- A QR-code/link will be shown, you can copy the link or scan the QR code with any WalletConnect compatible wallet and select an address & network on the walletConnect app.
- the selected address will show on the Metamask interface, select and confirm: the address will be added in the list of Metamask account
- You can use this account to sign tx or messages. When an action is required, it must be approved (signed) on the WalletConnect app (MetaMask has not direct access to the private key)

During initial testing, it appears that not all WalletConnect apps supports all WalletConnect requests, and some app do not seem to comply with the WalletConnect specifications. We try to follow [the official specification](https://docs.walletconnect.com/json-rpc-api-methods/ethereum) but more work may be required to improve support.

## Pre-Merge Checklist


### + If there are functional changes:
